src/App.tsx

```diff
@@ -1,24 +1,170 @@
+      .catch(() => showError(Errors.Add))
+      .finally(() => setTempTodo(null));
+  };
+
+  const deleteAllHandler = () => {
+    const completed = todos.filter(todo => todo.completed).map(todo => todo.id);
+
+    setProcessingTodos(completed);
+
+    Promise.all(todos.filter(todo => todo.completed)
+      .map(todo => deleteTodo(todo.id)))
+      .then(() => setTodos(currentTodos => {
+        return currentTodos.filter(todo => !todo.completed);
+      }))
+      .catch(() => showError(Errors.Delete))
+      .finally(() => setProcessingTodos([]));
+  };
+
+  const toggleAllHandler = () => {
+    const toggleTodos = (arr: Todo[], condition: boolean) => {
+      Promise.allSettled(arr.map(todo => {
+        return updateTodo(todo.id, { completed: condition });
+      }))
+        .then((results) => {
+          results.forEach(res => {
+            if (res.status === 'fulfilled') {
+              setTodos((current) => {
+                return current.map(todo => {
+                  if (todo.id === res.value.id) {
+                    return { ...todo, completed: condition };
+                  }
+
+                  return todo;
+                });
+              });
+            } else {
+              showError(Errors.Update);
+            }
+          });
+        })
+        .finally(() => setProcessingTodos([]));
+    };
+
+    if (todos.some(todo => !todo.completed)) {
+      const uncompleted = todos.filter(todo => !todo.completed);
+
+      setProcessingTodos(uncompleted.map(todo => todo.id));
+
+      toggleTodos(uncompleted, true);
+    } else {
+      setProcessingTodos(todos.map(todo => todo.id));
+
+      toggleTodos(todos, false);
+    }
+  };
+
+  useEffect(() => {
+    getTodos(USER_ID)
+      .then((setTodos))
+      .catch(() => {
+        showError(Errors.Load);
+      });
+  }, []);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-add-and-delete#react-todo-app-add-and-delete">React Todo App - Add and Delete</a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <TodosContext.Provider value={{ todos, setTodos }}>
+      <div className="todoapp">
+        <h1 className="todoapp__title">todos</h1>
+
+        <div className="todoapp__content">
+          <Header
+            userId={USER_ID}
+            onAdd={addTodoHandler}
+            onToggleAll={toggleAllHandler}
+          />
+
+          <Main
+            todos={visibleTodos}
+            tempTodo={tempTodo}
+            processingTodos={processingTodos}
+          />
+
+          {!!todos.length && (
+            <Footer
+              filter={filter}
+              onFilter={(value) => setFilter(value)}
+              onClear={() => {
+                deleteAllHandler();
+              }}
```

Comment:
```suggestion
              onFilter={setFilter}
              onClear={deleteAllHandler}
```
Your `setFilter` is already a function that accepts such an argument - wrapper is redundant
It's fair for all similar cases